### ISSUE TITLE
[Bug] Start Time on X1C always shows as Jan 1 1970 with incorrect time

### ISSUE THREAD

**Issue by swampylee** [greghesp/ha-bambulab#172]
### Describe the bug

In the last 2 releases I've noticed the Start Time shown for any job on my X1C is dated 1st Jan 1970 and the time is also wrong.

![image](https://github.com/greghesp/ha-bambulab/assets/52431327/01b6877c-adea-494b-97d8-101ab62c92c9)

### To Reproduce

Nothing special, just start a job.

### Expected Behaviour

Should show the date / time of when the job started based on users location details.

### What device are you using?

X1C

### Firmware Version

 01.06.01.00

### Home Assistant Version

2023.7.3

### Component Version

Latest non-beta

**Comment by Jhutjens92** [comment]
Same issue here. If i can provide any more debugging/logs/information then let me know.

**Comment by ferropolis** [comment]
+1
Same here

**Comment by AdrianGarside** [comment]
This was never support on the P1 and they've recently aligned the mqtt output closer between the two so i wonder if they dropped this for the X1? @greghesp - is this no longer working for you as well?

**Comment by AdrianGarside** [comment]
Yes, we'd have to do that if it's only in the full data blob. We'd need to do it on the various ways the print can end too.

**Comment by fmeus** [comment]
I have been tinkering the code in models.py. I sort of got it working by replacing the following line of code

`    self.start_time = start_time(int(data.get("gcode_start_time", self.remaining_time)))`

with

```
    if data.get("gcode_start_time") is not None:
        self.start_time = start_time(int(data.get("gcode_start_time")))
```

Due to the change the way the latest FW sends data (only sending what is changed instead of everything), the 'old' method does not work anymore since mc_remaining_time and gcode_start_time are send within separate MQTT messages.

The only problem is that start time is not being cleared when a print has finished or has been cancelled. But that is more caused by my limited knowledge of Python.

**Comment by AdrianGarside** [comment]
Yes, we should do it pybambu. My separation of concerns goal is that pybamby exposes (largely but not necessarily entirely modified) printer state. So it can modify the content it exposes to workaround issues or generate new sensors that make sense to be exposed from this layer. And then the controller layer takes that and modifies the content in ways necessary for best presentation purposes (and keep all the home assistant conversion knowledge up in this layer). This case is one that could be done in either location but since we know at least one person is using the pybambu layer for another project fixing up inconsistent/broken output like this in the pybambu layer is preferable.

**Comment by AdrianGarside** [comment]
I believe I've fixed this in: https://github.com/greghesp/ha-bambulab/releases/tag/v1.4.13-dev3

**Comment by ferropolis** [comment]
Hey @AdrianGarside ,
I'm not certain, but I've noticed in the last three prints the starting time are not correct. Maybe this is part of the recently made changes?

![2023-09-22 09_14_56-Übersicht – Home Assistant](https://github.com/greghesp/ha-bambulab/assets/77990395/fb9b8d33-44d1-43c2-8de7-840862977637)

I used v1.5.0 of Bambu Lab integration.

**Comment by AdrianGarside** [comment]
Can you show the full state of the integration?

**Comment by Jhutjens92** [comment]
Seems to be working for me? 
![image](https://github.com/greghesp/ha-bambulab/assets/14293742/87ebd03d-80b0-4a0c-b953-d5c4c82c1138)

**Comment by ferropolis** [comment]
> Can you show the full state of the integration?

Did you mean this one? 
![Einstellungen – Home Assistant - 192 168 2 6](https://github.com/greghesp/ha-bambulab/assets/77990395/c41a10a5-075c-4980-9b2c-083ccad514ed)


> @ferropolis - I've just pushed a pre-release of 1.6.0. Could you try that version, and then upload a diagnostic file? https://github.com/greghesp/ha-bambulab/tree/develop#diagnostic-file

I try...

**Comment by ferropolis** [comment]
> @ferropolis - I've just pushed a pre-release of 1.6.0. Could you try that version, and then upload a diagnostic file? https://github.com/greghesp/ha-bambulab/tree/develop#diagnostic-file

It look better, nearly perfect. <3
[Einstellungen – Home Assistant - 192 168 2 6_V2](https://github.com/greghesp/ha-bambulab/assets/77990395/c9586c3b-4cc2-4585-bdf5-5376bdf216ad)

**Comment by fmeus** [comment]
@greghesp I am running into the same issue. The only way I can get the starting time to display is by forcing a refresh (using v1.6.0-dev1).

I think it is a logic issue.

On startup of the addon/HA self.remaining_time is set to 0 (and it will also be 0 at the end of a print job).

When a printjob starts  a message is received which contains gcode_start_time (but no mc_remaining_time, so this will still be zero).

Sequence id = 3123
`{"print":{"command":"push_status","gcode_file_prepare_percent":"100","gcode_start_time":"1695387329","gcode_state":"RUNNING","msg":1,"profile_id":"18919452","project_id":"18923458","sequence_id":"3123","subtask_id":"36333927","t_utc":1695387330071,"task_id":"36333926","total_layer_num":200}}`

Since there is a check/reset based on mc_remaining_time, start_time is being wiped.

The next message that is received does contain a value for mc_remaining_time which is non-zero, but since we already have wiped start_time it will remain Unavailable.

Sequence id = 3124
`{"print":{"ams_rfid_status":2,"command":"push_status","layer_num":0,"mc_percent":0,"mc_print_stage":"2","mc_remaining_time":128,"msg":1,"nozzle_target_temper":75.0,"sequence_id":"3124","t_utc":1695387331095}}`

**Comment by AdrianGarside** [comment]
Thanks. That's very helpful. I know how to fix this for good now.

**Comment by fmeus** [comment]
> Thanks. That's very helpful. I know how to fix this for good now.

That's a bold statement ;)

**Comment by AdrianGarside** [comment]
I have a question for folk. Currently when remaining time reaches zero we make the end time sensor go to unavailable. So once a print is finished the end time sensor goes unavailable. But my opinion is that it makes sense to keep it available. And the same logic for start time. Just because a print finished doesn't invalidate when it started/ended and both are still appropriate to show in HA.

**Comment by fmeus** [comment]
> I have a question for folk. Currently when remaining time reaches zero we make the end time sensor go to unavailable. So once a print is finished the end time sensor goes unavailable. But my opinion is that it makes sense to keep it available. And the same logic for start time. Just because a print finished doesn't invalidate when it started/ended and both are still appropriate to show in HA.

Also when a print has finished and you do a force refresh you will also see that the start time is still being sent from the printer. So I would prefer that the sensors in HA matches what is being sent from the printer.

When needed one could always create a template sensor to divert from this logic, but thats just me.

**Comment by AdrianGarside** [comment]
Great. That observation definitively cements my opinion we shouldn't be wiping out either start or end time just because the print finished.

**Comment by swampylee** [comment]
> I have a question for folk. Currently when remaining time reaches zero we make the end time sensor go to unavailable. So once a print is finished the end time sensor goes unavailable. But my opinion is that it makes sense to keep it available. And the same logic for start time. Just because a print finished doesn't invalidate when it started/ended and both are still appropriate to show in HA.

I agree - leave them until a new print session starts so we can see when the last print started/finished.

**Comment by AdrianGarside** [comment]
Added generated start time for P1P/S devices that don't have the value in the mqtt payload.

**Comment by AdrianGarside** [comment]
Fixed in 1.6.0 official release. Added bonus was I added a simulated version for P1P/S since it doesn’t natively have it in the mqtt payload.
